pkg/i2gw/ingress2gateway.go

```diff
@@ -198,3 +199,74 @@ func mergeGateways(gatewaResources []GatewayResources) (map[types.NamespacedName
+//     a unique Gateway.
+//
+// This behavior is likely to change after https://github.com/kubernetes-sigs/gateway-api/pull/1863 takes place.
+func MergeIRs(irs ...IR) (IR, field.ErrorList) {
+	mergedIRs := IR{
+		Gateways:        make(map[types.NamespacedName]GatewayContext),
+		GatewayClasses:  make(map[types.NamespacedName]gatewayv1.GatewayClass),
+		HTTPRoutes:      make(map[types.NamespacedName]HTTPRouteContext),
+		Services:        make(map[types.NamespacedName]*ServiceIR),
+		TLSRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TLSRoute),
+		TCPRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TCPRoute),
+		UDPRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.UDPRoute),
+		ReferenceGrants: make(map[types.NamespacedName]gatewayv1beta1.ReferenceGrant),
+	}
+	var errs field.ErrorList
+	mergedIRs.Gateways, errs = mergeGatewayContexts(irs)
+	if len(errs) > 0 {
+		return IR{}, errs
+	}
+	for _, gr := range irs {
+		maps.Copy(mergedIRs.GatewayClasses, gr.GatewayClasses)
+		maps.Copy(mergedIRs.HTTPRoutes, gr.HTTPRoutes)
+		maps.Copy(mergedIRs.TLSRoutes, gr.TLSRoutes)
+		maps.Copy(mergedIRs.TCPRoutes, gr.TCPRoutes)
+		maps.Copy(mergedIRs.UDPRoutes, gr.UDPRoutes)
+		maps.Copy(mergedIRs.ReferenceGrants, gr.ReferenceGrants)
+	}
+	return mergedIRs, errs
+}
+
+func mergeGatewayContexts(irs []IR) (map[types.NamespacedName]GatewayContext, field.ErrorList) {
+	newGatewayContexts := make(map[types.NamespacedName]GatewayContext)
+	errs := field.ErrorList{}
+
+	for _, ir := range irs {
+		for _, g := range ir.Gateways {
+			nn := types.NamespacedName{Namespace: g.Gateway.Namespace, Name: g.Gateway.Name}
+			if existingGatewayContext, ok := newGatewayContexts[nn]; ok {
+				g.Gateway.Spec.Listeners = append(g.Gateway.Spec.Listeners, existingGatewayContext.Gateway.Spec.Listeners...)
+				g.Gateway.Spec.Addresses = append(g.Gateway.Spec.Addresses, existingGatewayContext.Gateway.Spec.Addresses...)
+				g.GatewayIR = mergedGatewayIR(g.GatewayIR, existingGatewayContext.GatewayIR)
+			}
+			newGatewayContexts[nn] = GatewayContext{Gateway: g.Gateway}
+			// 64 is the maximum number of listeners a Gateway can have
+			if len(g.Spec.Listeners) > 64 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("listeners")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 64 listeners"))
+			}
+			// 16 is the maximum number of addresses a Gateway can have
+			if len(g.Spec.Addresses) > 16 {
+				fieldPath := field.NewPath(fmt.Sprintf("%s/%s", nn.Namespace, nn.Name)).Child("spec").Child("addresses")
+				errs = append(errs, field.Invalid(fieldPath, g, "error while merging gateway listeners: a gateway cannot have more than 16 addresses"))
+			}
+		}
+	}
+	return newGatewayContexts, errs
+}
+
+func mergedGatewayIR(current, existing GatewayIR) GatewayIR {
```

Comment:
I dont like the specific need to call merge here... but I understand its a pain to solve now. Please add a todo and an issue on the repo so we can get this PR merged but we should revisit this soon.

Reply:
Done, added an issue to track this